src/features/performance/Controls.tsx

```diff
@@ -0,0 +1,118 @@
+import * as React from 'react';
+import {useDispatch} from 'react-redux';
+import {useAppSelector} from '../../hooks';
+import {setDataflowColoring, selectedColoringSelector} from '../dataflow/colorSlice';
+import {Values} from '../dataflow/pulsesSlice';
+import {setTimingRange} from '../dataflow/rangeSlice';
+import {selectedValuesSelector} from '../dataflow/selectionSlice';
+import './Controls.css';
+
+export function Controls() {
+  const dispatch = useDispatch();
+  const coloring: string = useAppSelector(selectedColoringSelector);
+  const values = useAppSelector(selectedValuesSelector);
+
+  const toggle = (e) => dispatch(setDataflowColoring((e.target as HTMLInputElement).value));
+
+  return (
+    <div>
+      <form>
+        <p>Color dataflow graph by:</p>
+        <input type="radio" value="type" name="dataflow_color" onChange={toggle} checked={coloring === 'type'} />
+        Node Type
+        <input type="radio" value="time" name="dataflow_color" onChange={toggle} defaultChecked={coloring === 'time'} />
+        Time
+      </form>
+      {values !== null && (
+        <Slider
+          values={values}
+          onChange={({min, max}: {min: number; max: number}) => dispatch(setTimingRange({min, max}))}
+        />
+      )}
+    </div>
+  );
+}
+
+interface SliderProps {
+  values: Values;
+  onChange: ({min, max}: {min: number; max: number}) => void;
+}
+const Slider: React.FC<SliderProps> = ({values, onChange}) => {
+  const time = Object.entries(values).map((v) => {
+    return v[1]['value'].time as number;
+  });
+  const max = Math.ceil(Math.max(...time) * 100) / 100;
+  const min = Math.floor(Math.min(...time) * 100) / 100;
+  const [minVal, setMinVal] = React.useState(min);
+  const [maxVal, setMaxVal] = React.useState(max);
+  const minValRef = React.useRef<HTMLInputElement>(null);
+  const maxValRef = React.useRef<HTMLInputElement>(null);
+  const range = React.useRef<HTMLDivElement>(null);
+  const getPercent = React.useCallback((value: number) => Math.round(((value - min) / (max - min)) * 100), [min, max]);
+
+  React.useEffect(() => {
+    if (maxValRef.current) {
+      const minPercent = getPercent(minVal);
+      const maxPercent = getPercent(+maxValRef.current.value);
+
+      if (range.current) {
+        range.current.style.left = `${minPercent}%`;
+        range.current.style.width = `${maxPercent - minPercent}%`;
+      }
+    }
+  }, [minVal, getPercent]);
+
+  React.useEffect(() => {
+    if (minValRef.current) {
+      const minPercent = getPercent(+minValRef.current.value);
+      const maxPercent = getPercent(maxVal);
```

Comment:
slider glitch when selecting pulses with smaller max than previous selected pulse. 
<img width="871" alt="image" src="https://user-images.githubusercontent.com/35775163/203488925-ed87af2c-d921-4d0e-b07d-074b3e7563eb.png">

Reply:
I tried to resolve this but haven't finished yet.

Reply:
Good catch. Got it.